### PR TITLE
Listenable.merge

### PR DESCRIPTION
**Pull request by Hixie**
Sometimes you have several listenables, but you want to hand them to an
API (e.g. CustomPainter) that only expects one.

### CODE REVIEW

**Comment by Hixie** [comment]
@abarth I'm posting this for your review. If you like it I'll write a test before submitting.

I'm mostly looking for your feedback on the API being a static on Listenable and the class it returns being private. It requires merging listenable.dart and change_notifier.dart (or having a two-way dependency), but it does avoid having yet another class in the docs.

File: packages/flutter/lib/src/foundation/change_notifier.dart [hunk 1]
```diff
@@ -6,7 +6,28 @@ import 'package:meta/meta.dart';
...
+
+  /// Remove a previously registered closure from the list of closures that the
+  /// object notifies.
+  void removeListener(VoidCallback listener);
+
+  /// Return a [Listenable] that triggers when any of the given [Listenable]s
+  /// themselves trigger.
+  ///
+  /// The list must not be changed after being this method has been called.
+  static Listenable merge(List<Listenable> listenables) {
```
Comment (abarth) [inline packages/flutter/lib/src/foundation/change_notifier.dart]: Looks reasonable, but I'd make this a factory constructor.

Comment (Hixie) [inline packages/flutter/lib/src/foundation/change_notifier.dart]: Seems legit.

**Comment by Hixie** [comment]
Ok, this is ready for review now.

cc @HansMuller

File: packages/flutter/lib/src/foundation/change_notifier.dart [hunk 2]
```diff
@@ -6,7 +6,26 @@ import 'package:meta/meta.dart';
...
+/// An object that maintains a list of listeners.
+abstract class Listenable {
+  /// Abstract const constructor. This constructor enables subclasses to provide
+  /// const constructors so that they can be used in const expressions.
+  const Listenable();
+
+  /// Return a [Listenable] that triggers when any of the given [Listenable]s
+  /// themselves trigger.
+  ///
+  /// The list must not be changed after being this method has been called.
```
Comment (Pacane) [inline packages/flutter/lib/src/foundation/change_notifier.dart]: Reword

File: packages/flutter/lib/src/foundation/change_notifier.dart [hunk 3]
```diff
@@ -69,3 +88,19 @@ class ChangeNotifier extends Listenable {
...
+
+  final List<Listenable> _children;
+
+  @override
+  void dispose() {
+    for (Listenable child in _children)
+      child.removeListener(notifyListeners);
+    super.dispose();
+  }
+}
```
Comment (Pacane) [inline packages/flutter/lib/src/foundation/change_notifier.dart]: dartfmt

File: packages/flutter/test/foundation/change_notifier_test.dart [hunk 4]
```diff
@@ -113,4 +113,28 @@ void main() {
...
+    final List<String> log = <String>[];
+
+    final Listenable merged = new Listenable.merge(<Listenable>[source1, source2]);
+    final VoidCallback listener = () { log.add('listener'); };
+
+    merged.addListener(listener);
+    source1.notify();
+    source2.notify();
+    source3.notify();
+    expect(log, equals(<String>['listener', 'listener']));
```
Comment (Pacane) [inline packages/flutter/test/foundation/change_notifier_test.dart]: `equals` is the default matcher.

File: packages/flutter/lib/src/foundation/change_notifier.dart (hunk 2 above)
Comment (Hixie) [inline packages/flutter/lib/src/foundation/change_notifier.dart]: Oh wow yeah, what write I was incomprehensive was.

File: packages/flutter/test/foundation/change_notifier_test.dart (hunk 4 above)
Comment (Hixie) [inline packages/flutter/test/foundation/change_notifier_test.dart]: ok, removed the equals() from everywhere in this file.

**Comment by Hixie** [comment]
PTAL.

File: packages/flutter/lib/src/foundation/change_notifier.dart (hunk 3 above)
Comment (Pacane) [inline packages/flutter/lib/src/foundation/change_notifier.dart]: Well maybe you want to add an empty line at the eof then.

Comment (Hixie) [inline packages/flutter/lib/src/foundation/change_notifier.dart]: If you insist. :-)

**Review by Pacane — APPROVED** [review]

**Review by chinmaygarde — APPROVED** [review]

**Comment by Hixie** [comment]
I'm going to assume that using the github approval tool is equivalent to an explicit "LGTM". Thanks. :-)

**Comment by chinmaygarde** [comment]
Oops. I usually type in LGTM as well as the approval.

**Pull request merged**